File: Libs/ConfigurationLib/src/configlib/XMLArchiverIn.java
package configlib;

import com.sun.org.apache.xerces.internal.parsers.DOMParser;
import configlib.Deserializable;
import configlib.exception.FormatException;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

/**
 *
 * @author Anton Gravestam
 */
public class XMLArchiverIn implements ArchiverInOut
{

    private InputStream is;
    private Document doc;
    private Node currentNode;
    private int currentElement =  -1;
    private Stack<Node> nodeStack = new Stack<Node>();

    private SerializableCompositeFactory factory = new SerializableCompositeFactory();

    public XMLArchiverIn(InputStream is) throws FormatException
    {
        try
        {
            this.is = is;
            DOMParser parser = new DOMParser();
            
            // Parse the document.
            parser.parse(new InputSource(is));
            // Obtain the document.
            doc = parser.getDocument();
            
            currentNode = doc;
        }
        catch (SAXException ex)
        {
            throw new FormatException("Caused by underlying SAXException: " + ex.getMessage());
        }
        catch (IOException ex)
        {
            throw new FormatException("Caused by underlying IOException: " + ex.getMessage());
        }
    }
    public XMLArchiverIn(InputStream is, String rootNode) throws FormatException
    {
        try
        {
            this.is = is;
            DOMParser parser = new DOMParser();

            // Parse the document.
            parser.parse(new InputSource(is));
            // Obtain the document.
            doc = parser.getDocument();

            currentNode = doc;
            currentNode = getNode(rootNode);
        }
        catch (SAXException ex)
        {
            throw new FormatException("Caused by underlying SAXException: " + ex.getMessage());
        }
        catch (IOException ex)
        {
            throw new FormatException("Caused by underlying IOException: " + ex.getMessage());
        }
    }

    public byte getByte(String name)
    {
        return Byte.parseByte(getValue(name));
        
    }

    public int getInt(String name)
    {
        return Integer.parseInt(getValue(name));
    }

    public short getShort(String name)
    {
        return Byte.parseByte(getValue(name));
    }

    public long getLong(String name)
    {
        return Long.parseLong(getValue(name));
    }

    public float getFloat(String name)
    {
        return Float.parseFloat(getValue(name));
    }

    public double getDouble(String name)
    {
        return Double.parseDouble(getValue(name));
    }

    public String getString(String name)
    {
        String ret = getValue(name);
        if(ret == null)
        {
            ret = "";
        }
        return ret;
    }

//    public void getDeserializable(String name, Deserializable des)
//    {
//        nodeStack.push(currentNode);
//        currentNode = getNode(name);
//        des.desirialize(this);
//        currentNode = nodeStack.pop();
//    }
    public int getNrElements(String name)
    {
        nodeStack.push(currentNode);
        currentNode = getNode(name);
        NodeList nodes = currentNode.getChildNodes();
        int size = 0;
        for(int i = 0 ; i < nodes.getLength(); i++)
        {
            if(nodes.item(i).getNodeName().equals("element"))
            {
                size++;                
            }
        }
        currentNode = nodeStack.pop();
        return size; 
        
    }
//    public Deserializable getElement(String name, int i, Deserializable deserializable)
//    {
//        try
//        {
//            nodeStack.push(currentNode);
//            currentNode = getNode(name);
//            nodeStack.push(currentNode);
//            currentNode = getCurrentElement(i);
//
//            //Deserializable deserializable = (Deserializable) cls.getConstructor(null).newInstance();
//
//            deserializable.desirialize(this);
//
//            currentNode = nodeStack.pop();
//            currentNode = nodeStack.pop();
//
//            return deserializable;
//        }
//        catch (Exception ex)
//        {
//           return null;
//        }
//
//
//
//
//
//    }
    private Node getCurrentElement(int n)
    {
        
        int ittElement = 0;
        NodeList nodes = currentNode.getChildNodes();
        for(int i = 0 ; i < nodes.getLength(); i++)
        {
            if(nodes.item(i).getNodeName().equals("element") && n == ittElement)
            {
                return nodes.item(i);
            }
            else if(nodes.item(i).getNodeName().equals("element"))
            {
                ittElement++;                
            }
        }
        
        return null;
        
    }

    private Node getNode(String name)
    {
        NodeList nodes = currentNode.getChildNodes();
        for(int i = 0 ; i < nodes.getLength(); i++)
        {
            if(nodes.item(i).getNodeName().equals(name))
            {
                return nodes.item(i);
            }
        }
        return null;
        
    }

    private String getValue(String name)
    {
        NodeList nodes = currentNode.getChildNodes();
        for(int i = 0 ; i < nodes.getLength(); i++)
        {
            if(nodes.item(i).getNodeName().equals(name))
            {
                return nodes.item(i).getTextContent();
            }
            
        }
        return null;
    }

//    public void getList(String name, AddElementCallback addElementCallback, Class<? extends Deserializable> cls)
//    {
//        nodeStack.push(currentNode);
//        currentNode = getNode(name);
//        NodeList nodes = currentNode.getChildNodes();
//        for(int i = 0 ; i < nodes.getLength(); i++)
//        {
//            if(nodes.item(i).getNodeName().equals("element"))
//            {
//                try
//                {
//                    Deserializable deserializable = (Deserializable) cls.getConstructor(null).newInstance();
//                    getDeserializable("element", deserializable);
//                    addElementCallback.addElement(deserializable);
//
//                }
//                catch (NoSuchMethodException ex)
//                {
//                    Logger.getLogger(XMLArchiverIn.class.getName()).log(Level.SEVERE, null, ex);
//                }
//                catch (SecurityException ex)
//                {
//                    Logger.getLogger(XMLArchiverIn.class.getName()).log(Level.SEVERE, null, ex);
//                }
//                catch (InstantiationException ex)
//                {
//                    Logger.getLogger(XMLArchiverIn.class.getName()).log(Level.SEVERE, null, ex);
//                }
//                catch (IllegalAccessException ex)
//                {
//                    Logger.getLogger(XMLArchiverIn.class.getName()).log(Level.SEVERE, null, ex);
//                }
//                catch (IllegalArgumentException ex)
//                {
//                    Logger.getLogger(XMLArchiverIn.class.getName()).log(Level.SEVERE, null, ex);
//                }
//                catch (InvocationTargetException ex)
//                {
//                    Logger.getLogger(XMLArchiverIn.class.getName()).log(Level.SEVERE, null, ex);
//                }
//            }
//        }
//
//
//        currentNode = nodeStack.pop();
//
//    }

    public boolean getBoolean(String name)
    {
        return Boolean.parseBoolean(getValue(name));
    }

    public int inout(String name, int v) throws IOException
    {
        return getInt(name);
    }

    public long inout(String name, long v) throws IOException
    {
        return getLong(name);
    }

    public byte inout(String name, byte v) throws IOException
    {
        return getByte(name);
    }

    public short inout(String name, short v) throws IOException
    {
        return getShort(name);
    }

    public float inout(String name, float v) throws IOException
    {
        return getFloat(name);
    }

    public boolean inout(String name, boolean v) throws IOException
    {
        return getBoolean(name);
    }

    public String inout(String name, String v) throws IOException
    {
        return getString(name);
    }

    public double inout(String name, double v) throws IOException
    {
        return getDouble(name);
    }

    public Serializable inout(String name, Serializable v) throws IOException
    {
        nodeStack.push(currentNode);
        currentNode = getNode(name);
        String type = currentNode.getAttributes().getNamedItem("type").getNodeValue();
        Serializable newElem = factory.create(type);
        newElem.serialize(this);
        currentNode = nodeStack.pop();
        return newElem;
    }

    public List inout(String name, List v) throws IOException
    {
        int size = getNrElements(name);
        for (int i = 0; i < size; i++)
        {
            v.add(getElement(name, i));
        }
        return v;
    }
    public Serializable getElement(String name, int i)
    {
        try
        {
            nodeStack.push(currentNode);
            currentNode = getNode(name);
            nodeStack.push(currentNode);
            currentNode = getCurrentElement(i);
            String type = currentNode.getAttributes().getNamedItem("type").getNodeValue();

            Serializable newElem = factory.create(type);

            newElem.serialize(this);

            currentNode = nodeStack.pop();
            currentNode = nodeStack.pop();

            return newElem;
        }
        catch (Exception ex)
        {
           return null;
        }
    }

    public boolean remove(Object o)
    {
        return factory.remove(o);
    }

    public boolean add(SerializableFactory e)
    {
        return factory.add(e);
    }

    public List<Integer> inoutIntegerList(String name, List<Integer> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Long> inoutLongList(String name, List<Long> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Byte> inoutByteList(String name, List<Byte> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Short> inoutShortList(String name, List<Short> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Float> inoutFloatList(String name, List<Float> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Boolean> inoutBooleanList(String name, List<Boolean> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<String> inoutStringList(String name, List<String> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Double> inoutDoubleList(String name, List<Double> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List inoutSerializableList(String name, List v) throws IOException
    {
        int size = getNrElements(name);
        for (int i = 0; i < size; i++)
        {
            v.add(getElement(name, i));
        }
        return v;
    }



}

File: C#/Ops/Archiver/XMLArchiverIn.cs
///////////////////////////////////////////////////////////
//  XMLArchiverIn.cs
//  Implementation of the Class XMLArchiverIn
//  Created on:      12-nov-2011 21:02:41
//  Author:
///////////////////////////////////////////////////////////

using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Xml;

namespace Ops
{
	public class XMLArchiverIn : IArchiverInOut 
    {
        private const NumberStyles numberStyles = NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint | NumberStyles.AllowExponent | NumberStyles.AllowLeadingWhite | NumberStyles.AllowTrailingWhite;
        private NumberFormatInfo numberFormatInfo = new NumberFormatInfo() { NumberDecimalSeparator = "." };
		private XmlNode currentNode;
        private XmlDocument doc = new XmlDocument();
		private SerializableCompositeFactory compositeFactory = new SerializableCompositeFactory();
		private Stack<XmlNode> nodeStack = new Stack<XmlNode>();

        public XMLArchiverIn(Stream s)
        {
            try
            {
                // Load the XML data from a file.
                doc.Load(s);                      

                currentNode = doc;
             } 
             catch(XmlException xmlEx)
             {
                throw new FormatException("Caused by underlying XML exception: " + xmlEx.Message);
             }
             catch(Exception ex)
             {
                throw new FormatException("Caused by underlying generic exception: " + ex.Message);
             }
             finally
             {
                // Finalize here.
             }
        }

        public XMLArchiverIn(Stream s, string rootNode)
        {
            try
            {
                // Load the XML data from a file.
                doc.Load(s);                      

                currentNode = doc;
                currentNode = GetNode(rootNode);
            }
             catch(XmlException xmlEx)
             {
                throw new FormatException("Caused by underlying XML exception: " + xmlEx.Message);
             }
             catch(Exception ex)
             {
                throw new FormatException("Caused by underlying generic exception: " + ex.Message);
             }
             finally
             {
                // Finalize here.
             }
        }

        public bool IsOut()
        {
            return false;
        }

        public int GetNrElements(string name)
        {
            nodeStack.Push(currentNode);
            try
            {
                currentNode = GetNode(name);
                if (currentNode == null)
                {
                    return 0;
                }
                XmlNodeList nodes = currentNode.ChildNodes;
                int size = 0;
                foreach (XmlNode node in nodes)
                {
                    if (node.Name.Equals("element"))
                    {
                        size++;
                    }
                }
                return size;
            }
            finally
            {
                currentNode = nodeStack.Pop();
            }
        }

        private XmlNode GetCurrentElement(int n)
        {
            int ittElement = 0;
            XmlNodeList nodes = currentNode.ChildNodes;
            foreach (XmlNode node in nodes)
            {
                if (node.Name.Equals("element") && n == ittElement)
                {
                    return node;
                } 
                else
                {
                    if (node.Name.Equals("element"))
                    {
                        ittElement++;
                    }
                }
            }
            return null;
        }

        private XmlNode GetNode(string name)
        {
            XmlNodeList nodes = currentNode.ChildNodes;
            foreach (XmlNode node in nodes)
            {
                if (node.Name.Equals(name))
                {
                    return node;
                }
            }
            return null;
        }

        private string GetValue(string name)
        {
            XmlNodeList nodes = currentNode.ChildNodes;
            foreach (XmlNode node in nodes)
            {
                if (node.Name.Equals(name))
                {
                    return node.InnerText.Trim();
                }
            }
            return null;
        }

        // Read a "bool" value with name "name". If not found return the input value "v".
        public bool Inout(string name, bool v)
        {
            bool ret;
            if (bool.TryParse(GetValue(name), out ret))
            {
                return ret;
            }
            else
            {
                return v;
            }
        }

        // Read a "byte" value with name "name". If not found return the input value "v".
        public byte Inout(string name, byte v)
        {
            byte ret;
            if (byte.TryParse(GetValue(name), out ret))
            {
                return ret;
            }
            else
            {
                return v;
            }
        }

        // Read a "double" value with name "name". If not found return the input value "v".
        public double Inout(string name, double v)
        {
            double ret;

            // A specified NumberFormatInfo ensures that values, regardless of regional format settings in Windows, 
            // always are read with "." as decimal separator
            if (double.TryParse(GetValue(name), numberStyles, numberFormatInfo, out ret))
            {
                return ret;
            }
            else
            {
                return v;
            }
        }

        // Read a "float" value with name "name". If not found return the input value "v".
        public float Inout(string name, float v)
        {
            float ret;

            // A specified NumberFormatInfo ensures that values, regardless of regional format settings in Windows, 
            // always are read with "." as decimal separator
            if (float.TryParse(GetValue(name), numberStyles, numberFormatInfo, out ret))
            {
                return ret;
            }
            else
            {
                return v;
            }
        }

        // Read an "int" value with name "name". If not found return the input value "v".
        public int Inout(string name, int v) 
        {
            int ret;
            if (int.TryParse(GetValue(name), out ret))
            {
                return ret;
            }
            else
            {
                return v;
            }
        }

        // Read a "long" value with name "name". If not found return the input value "v".
        public long Inout(string name, long v)
        {
            long ret;
            if (long.TryParse(GetValue(name), out ret))
            {
                return ret;
            }
            else
            {
                return v;
            }
        }

        // Read a "short" value with name "name". If not found return the input value "v".
        public short Inout(string name, short v)
        {
            short ret;
            if (short.TryParse(GetValue(name), out ret))
            {
                return ret;
            }
            else
            {
                return v;
            }
        }

        // Read a "string" value with name "name". If not found return an empty string.
        public string Inout(string name, string v)
        {
            string ret = GetValue(name);
            if (ret == null)
            {
                ret = "";
            }
            return ret;
        }

        // Read a "ISerializable" item with name "name". 
        public ISerializable Inout(string name, ISerializable v) 
        {
            nodeStack.Push(currentNode);
            currentNode = GetNode(name);
            string type = currentNode.Attributes.GetNamedItem("type").Value;
            ISerializable newElem = compositeFactory.Create(type);
            newElem.Serialize(this);
            currentNode = nodeStack.Pop();
            return newElem;
        }

        public ISerializable GetElement(string name, int i)
        {
            try
            {
                nodeStack.Push(currentNode);
                currentNode = GetNode(name);
                nodeStack.Push(currentNode);
                currentNode = GetCurrentElement(i);
                string type = currentNode.Attributes.GetNamedItem("type").Value;

                ISerializable newElem = compositeFactory.Create(type);

                newElem.Serialize(this);

                currentNode = nodeStack.Pop();
                currentNode = nodeStack.Pop();

                return newElem;
            } 
            catch (Exception)
            {
                return null;
            }
        }

        public void Remove(ISerializableFactory item)
        {
            compositeFactory.Remove(item);
        }

        public void Add(ISerializableFactory item)
        {
            compositeFactory.Add(item);
        }

        public List<int> InoutIntegerList(string name, List<int> v)
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        public List<long> InoutLongList(string name, List<long> v) 
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        public List<Byte> InoutByteList(string name, List<Byte> v) 
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        public List<short> InoutShortList(string name, List<short> v) 
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        public List<float> InoutFloatList(string name, List<float> v) 
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        public List<bool> InoutBooleanList(string name, List<bool> v) 
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        public List<string> InoutStringList(string name, List<string> v) 
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        public List<double> InoutDoubleList(string name, List<double> v) 
        {
            throw new System.NotSupportedException("Not supported yet.");
        }

        // NB! We assume that object "v" is a generic list.
        public IList InoutSerializableList(string name, IList v)
        {
            Type type = v.GetType().GetGenericArguments()[0];
            IList list = (IList)Activator.CreateInstance((typeof(List<>).MakeGenericType(type)));
            int size = GetNrElements(name);
            for (int i = 0; i < size; i++)
            {
                list.Add((ISerializable)GetElement(name, i));
            }
            return list;
        }

        // NB! We assume that object "v" is a generic list.
        public IList InoutSerializableList<T>(string name, IList v)
        {
            return InoutSerializableList(name, v);
        }

    }

}